docs/design.md

```diff
@@ -0,0 +1,88 @@
+way to share container images among pods/containers running on the same physical node. But it's time consuming to
+download, verify, decrypt and extract container images for each pod/container. So the `image-rs` crate will explore
+technologies and innovations to reduce image preparation time. Data decryption and decompression acceleration libraries
+may be used to speed up image decryption and decompression. Image data on demand loading may also be adopted to reduce
+container startup time.
+
+### Goals
+ * Support pulling OCI/docker v2 images from remote registries compatible with the OCI distribution spec.
+ * Support downloading OCI images from local [OCI layout compatiable](https://github.com/opencontainers/image-spec/blob/main/image-layout.md) storage
+ * Support OCI image decryption
+ * Support OCI image signature verification
+ * Support seamlessly integration with Kata Agent and Attestation Agent
+ * Explore image decryption and decompression acceleration technologies
+ * Explore image on demand loading technologies
+
+
+### Non-Goals
+ * Image push, content discovery/management operations for remote registries
+ * Image encryption and signing
+ * Image export to local storage
+ * Image import from other service daemon or other local storage format
+
+
+### Existing Solutions
+ * [containers-image-proxy-rs](https://github.com/containers/containers-image-proxy-rs): a rust bindings for skopeo,
+   which doesn't meet our requirements of small TCB and lightweight.
+ * [oci-distribution](https://github.com/krustlet/oci-distribution): a crate from
+   [krustlet](https://github.com/krustlet/krustlet) to run wasm application with K8S. Since wasm is platform agnostic
+   and small code size, `oci-distribution` does not support manifest lists for different platforms and internal data
+   struct is not friend to generic container workload.
+ * [oci-registry-client](https://github.com/ecarrara/oci-registry-client): a very simple client for OCI image registries
+   with limited features and latest commit is one and half years ago.
+
+Above projects focus on image distribution and do not support image unpack and storage management features.
+
+
+## Architecture
+The following diagram demonstrates the overall architecture and how to
+integrate with Kata Agent:
+![Architecture](images/architecture.png)
+
+`image-rs` provides the API which mainly cover image pulling and management.
+
+For image pulling, it will follow OCI distribution and image spec to
+interactive with remote registries to download container image metadata
+and blob data, these info will be consumed by the image management module.
+After container image is downloaded, `image-rs` will talk with remote
+attestation service to verify the image signature and decrypt the image
+if needed. After image is unpacked, the image format module will handle
+the format conversion between oci v1 and docker v2.
```

Comment:
Is local caching relevant at all? 
Or is it the case that after everything is decrypted and unpacked the original blobs will be destroyed?

Reply:
To be honest, we are not very clear about this yet:(
For most cases, we do not care about local caching. But not sure whether it will be needed for image data sharing.

Reply:
Caching might be good if you have more than one of the same container inside a pod. We definitely don't want to download a container more than once if it can be avoided.

Reply:
If caching is to be supported, is additional security needed for it? And how is it managed?

Reply:
> 

We should support the case where an image is shared by multiple containers within a pod.
On the other handle, it's challenge to share container images among pods, even not support by current hardware yet.
So let's leave the hard part to the future.

Reply:
Yes, will add this info.